ksp_plugin/pile_up.cpp

```diff
@@ -416,6 +406,68 @@ void PileUp::AppendToPart(DiscreteTrajectory<Barycentric>::Iterator it) const {
   }
 }
 
+DegreesOfFreedom<Barycentric> PileUp::RecomputeFromParts(
+    std::list<not_null<Part*>> const& parts) {
+  // First compute the overall mass and centre of mass of the pile-up.  Also
+  // compute the overall force applied to it.
+  // TODO(phl): We assume that forces are applied at the centre of mass of the
+  // pile-up, but they are really applied at some unknown point of the parts, so
+  // this introduces a torque.
+  Mass pile_up_mass;
+  Vector<Force, Barycentric> pile_up_intrinsic_force;
+  BarycentreCalculator<DegreesOfFreedom<Barycentric>, Mass> calculator;
+  for (not_null<Part*> const part : parts) {
+    pile_up_intrinsic_force += part->intrinsic_force();
+    calculator.Add(part->degrees_of_freedom(), part->inertia_tensor().mass());
+  }
+  pile_up_mass = calculator.weight();
+  DegreesOfFreedom<Barycentric> const pile_up_barycentre = calculator.Get();
+
+  // Then compute the inertia tensor and the angular momentum of the pile-up.
+  InertiaTensor<RigidPileUp> pile_up_inertia_tensor;
+  Bivector<AngularMomentum, RigidPileUp> pile_up_angular_momentum;
+  RigidMotion<RigidPileUp, Barycentric> const pile_up_to_barycentric(
+      RigidTransformation<RigidPileUp, Barycentric>(
+          RigidPileUp::origin,
+          pile_up_barycentre.position(),
+          Identity<RigidPileUp, Barycentric>().Forget()),
+      AngularVelocity<Barycentric>{},
+      pile_up_barycentre.velocity());
+  RigidMotion<Barycentric, RigidPileUp> const barycentric_to_pile_up =
+      pile_up_to_barycentric.Inverse();
+  for (not_null<Part*> const part : parts) {
+    RigidMotion<RigidPart, Barycentric> const& part_to_barycentric =
+        part->rigid_motion();
+    RigidMotion<RigidPart, RigidPileUp> const part_to_pile_up =
+        barycentric_to_pile_up * part_to_barycentric;
+
+    // NOTE(phl): The following call reads like the tensor "transforms" the
+    // rigid motion.  Improve this API.
+    InertiaTensor<RigidPileUp> const& part_inertia_tensor =
+        part->inertia_tensor().Transform(
+            part_to_pile_up.rigid_transformation());
+    pile_up_inertia_tensor += part_inertia_tensor;
+
+    Bivector<AngularMomentum, RigidPileUp> const part_angular_momentum =
+        Anticommutator(
+            part_inertia_tensor,
+            part_to_pile_up.Inverse().angular_velocity_of_to_frame());
+    DegreesOfFreedom<RigidPileUp> const part_degrees_of_freedom =
+        part_to_pile_up({RigidPart::origin, Velocity<RigidPart>{}});
+    pile_up_angular_momentum +=
+        part_angular_momentum +
+        Wedge(part_degrees_of_freedom.position() - RigidPileUp::origin,
+              part->inertia_tensor().mass() *
+                  part_degrees_of_freedom.velocity()) *
+            Radian;
```

Comment:
Please open an issue to track moving the computation of the momenta from the motion and the inertia to some place that can be unit-tested.